src/v/http/probe.h

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include <cstdint>
+
+namespace http {
+
+/// Simple differential counting for requests
+struct diff_counter {
+    diff_counter() = default;
+    /// Count request start
+    void start() { _started++; }
+    /// Count request termination
+    void complete() { _completed++; }
+    /// Total number of completed requests
+    uint64_t total() const { return _completed; }
+    /// Number of active requests
+    uint64_t active() const {
+        if (_completed <= _started) {
+            return _started - _completed;
+        }
+        return 0;
+    }
+
+private:
+    uint64_t _started;
+    uint64_t _completed;
+};
+
+class client_probe {
+public:
+    enum class verb {
+        get,
+        put,
+        other,
+    };
+
+    client_probe() = default;
+
+    /// Register incomming traffic
+    void add_inbound_bytes(uint64_t bytes) { _in += bytes; }
+
+    /// Register outgoing traffic
+    void add_outbound_bytes(uint64_t bytes) { _out += bytes; }
+
+    /// RAII wrapper for per-request probe
+    struct subprobe {
+        verb v;
+        client_probe& probe;
+        subprobe(verb v, client_probe& p)
+          : v(v)
+          , probe(p) {
+            switch (v) {
+            case verb::get:
+                probe._get_requests.start();
+                break;
+            case verb::put:
+                probe._put_requests.start();
+                break;
+            default:
+                break;
+            }
+            probe._all_requests.start();
+        }
+        subprobe& operator=(const subprobe&) = delete;
+        subprobe& operator=(subprobe&&) = delete;
+        subprobe(const subprobe&) = delete;
+        subprobe(subprobe&&) = delete;
+        ~subprobe() {
+            switch (v) {
+            case verb::get:
+                probe._get_requests.complete();
+                break;
+            case verb::put:
+                probe._put_requests.complete();
+                break;
+            default:
+                break;
+            }
+            probe._all_requests.complete();
+        }
+    };
+
+    auto create_request_subprobe(verb v) { return subprobe(v, *this); }
```

Comment:
this is great!

in the future we might want `unique_ptr< pb > create_request_subprobe_ptr(..)` so it can be moved to a lambda capture too.